From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Wed, 17 Mar 2021 23:49:04 -0700
Subject: feat: Optionally make FindRequestManager to start search from a RFH

This patch adds the ability to search within a subtree of frames,
rather than always starting from the top-level WebContents.

Possible to upstream https://issues.chromium.org/issues/40755645

diff --git a/chrome/browser/ui/browser.cc b/chrome/browser/ui/browser.cc
index d22c3f134573a9ba91cedf24aa4dafc48332914f..9dcf7ebb4313e1dabf430651b41ff977d9f7d10b 100644
--- a/chrome/browser/ui/browser.cc
+++ b/chrome/browser/ui/browser.cc
@@ -2839,7 +2839,8 @@ void Browser::FindReply(WebContents* web_contents,
                         int number_of_matches,
                         const gfx::Rect& selection_rect,
                         int active_match_ordinal,
-                        bool final_update) {
+                        bool final_update,
+                        content::RenderFrameHost* target_rfh) {
   find_in_page::FindTabHelper* find_tab_helper =
       find_in_page::FindTabHelper::FromWebContents(web_contents);
   if (!find_tab_helper) {
diff --git a/chrome/browser/ui/browser.h b/chrome/browser/ui/browser.h
index 037f5ecdf3a4437c935b90c45ef29536e0bfa2e3..a4eb2ed0e88d74f5908ef4d2f1d99c68091cc51c 100644
--- a/chrome/browser/ui/browser.h
+++ b/chrome/browser/ui/browser.h
@@ -1021,7 +1021,8 @@ class Browser : public TabStripModelObserver,
                  int number_of_matches,
                  const gfx::Rect& selection_rect,
                  int active_match_ordinal,
-                 bool final_update) override;
+                 bool final_update,
+                 content::RenderFrameHost* target_rfh) override;
   void RequestPointerLock(content::WebContents* web_contents,
                           bool user_gesture,
                           bool last_unlocked_by_target) override;
diff --git a/content/browser/find_request_manager.cc b/content/browser/find_request_manager.cc
index 33adf341692624cc850607c3c238855e5035fdd3..7624bacd8d014099f24f43e6aa99939700f81166 100644
--- a/content/browser/find_request_manager.cc
+++ b/content/browser/find_request_manager.cc
@@ -211,6 +211,9 @@ class FindRequestManager::FrameObserver : public WebContentsObserver {
 
   void RenderFrameDeleted(RenderFrameHost* rfh) override {
     manager_->RemoveFrame(rfh);
+    if (manager_->GetTargetRenderFrameHost() == rfh) {
+      manager_->SetTargetRenderFrameHost(nullptr);
+    }
   }
 
   void RenderFrameHostStateChanged(
@@ -316,6 +319,10 @@ const int FindRequestManager::kInvalidId = -1;
 FindRequestManager::FindRequestManager(WebContentsImpl* web_contents)
     : contents_(web_contents) {}
 
+FindRequestManager::FindRequestManager(WebContentsImpl* web_contents,
+                                       RenderFrameHostImpl* rfh)
+    : contents_(web_contents), target_rfh_(rfh), target_rfh_assigned_(rfh != nullptr) {}
+
 FindRequestManager::~FindRequestManager() = default;
 
 void FindRequestManager::Find(int request_id,
@@ -509,8 +516,9 @@ void FindRequestManager::RemoveFrame(RenderFrameHost* rfh) {
   // is removed, we can target any queued requests to the focused frame or
   // primary main frame. However, if the primary main frame is removed we will
   // not have a valid RenderFrameHost to target for the request queue.
-  if (rfh->IsInPrimaryMainFrame())
+  if (rfh->IsInPrimaryMainFrame() || target_rfh_ == rfh) {
     find_request_queue_ = base::queue<FindRequest>();
+  }
 
   // Update the active match ordinal, since it may have changed.
   if (active_frame_ == rfh) {
@@ -701,6 +709,15 @@ void FindRequestManager::FindInternal(const FindRequest& request) {
   // This is an initial find operation.
   Reset(request);
 
+  if (target_rfh_) {
+    frame_observers_.push_back(
+        std::make_unique<FrameObserver>(contents_, this));
+    target_rfh_->ForEachRenderFrameHost([this](content::RenderFrameHost* rfh) {
+      AddFrame(rfh, false /* force */);
+    });
+    return;
+  }
+
   // Add and observe eligible RFHs in the WebContents. And, use
   // ForEachRenderFrameHostImpl instead of ForEachAddedFindInPageRenderFrameHost
   // because that calls CheckFrame() which will only be true if we've called
@@ -757,13 +774,20 @@ void FindRequestManager::NotifyFindReply(int request_id, bool final_update) {
   else
     last_reported_id_ = request_id;
 
-  contents_->NotifyFindReply(request_id, number_of_matches_, selection_rect_,
-                             active_match_ordinal_, final_update);
+  if (target_rfh_assigned_ && CheckFrame(target_rfh_)) {
+    contents_->NotifyFindReply(request_id, number_of_matches_, selection_rect_,
+                               active_match_ordinal_, final_update, target_rfh_);
+  } else if (!target_rfh_assigned_) {
+    contents_->NotifyFindReply(request_id, number_of_matches_, selection_rect_,
+                               active_match_ordinal_, final_update, nullptr);
+  }
 }
 
 RenderFrameHost* FindRequestManager::GetInitialFrame(bool forward) const {
   RenderFrameHost* rfh =
-      contents_->GetPrimaryFrameTree().root()->current_frame_host();
+      target_rfh_
+          ? target_rfh_.get()
+          : contents_->GetPrimaryFrameTree().root()->current_frame_host();
   if (!forward)
     rfh = GetDeepestLastChild(static_cast<RenderFrameHostImpl*>(rfh));
 
@@ -933,6 +957,14 @@ std::unique_ptr<FindInPageClient> FindRequestManager::CreateFindInPageClient(
   return std::make_unique<FindInPageClient>(this, rfh);
 }
 
+raw_ptr<RenderFrameHost> FindRequestManager::GetTargetRenderFrameHost() {
+  return target_rfh_;
+}
+
+void FindRequestManager::SetTargetRenderFrameHost(RenderFrameHost* rfh) {
+  target_rfh_ = static_cast<RenderFrameHostImpl*>(rfh);
+}
+
 #if BUILDFLAG(IS_ANDROID)
 void FindRequestManager::RemoveNearestFindResultPendingReply(
     RenderFrameHost* rfh) {
diff --git a/content/browser/find_request_manager.h b/content/browser/find_request_manager.h
index 1b6f53b4394fcc2719e56abfe650e0992b972883..e67c2daa497fa74fc359ce5f10897963332a3a6d 100644
--- a/content/browser/find_request_manager.h
+++ b/content/browser/find_request_manager.h
@@ -39,10 +39,10 @@ class WebContentsImpl;
 class FindRequestManager {
  public:
   explicit FindRequestManager(WebContentsImpl* web_contents);
-
+  FindRequestManager(WebContentsImpl* web_contents,
+                     RenderFrameHostImpl* target_rfh);
   FindRequestManager(const FindRequestManager&) = delete;
   FindRequestManager& operator=(const FindRequestManager&) = delete;
-
   ~FindRequestManager();
 
   // Initiates a find operation for |search_text| with the options specified in
@@ -129,6 +129,9 @@ class FindRequestManager {
     create_find_in_page_client_for_testing_ = create_func;
   }
 
+  raw_ptr<RenderFrameHost> GetTargetRenderFrameHost();
+  void SetTargetRenderFrameHost(RenderFrameHost* target_rfh);
+
  private:
   friend class FindRequestManagerFencedFrameTest;
 
@@ -308,6 +311,10 @@ class FindRequestManager {
   // WebContentses within it will be searched.
   const raw_ptr<WebContentsImpl> contents_;
 
+  // When provided only frames in |target_rfh_| will be searched.
+  raw_ptr<RenderFrameHostImpl> target_rfh_ = nullptr;
+  bool target_rfh_assigned_ = false;
+
   // The request ID of the initial find request in the current find-in-page
   // session, which uniquely identifies this session. Request IDs are included
   // in all find-related IPCs, which allows reply IPCs containing results from
diff --git a/content/browser/web_contents/web_contents_impl.cc b/content/browser/web_contents/web_contents_impl.cc
index 5b9daf8e31da1ebef7d154cb2a970f7d68566a12..3e2a46319e2228c82fd56909a716a63e1504c482 100644
--- a/content/browser/web_contents/web_contents_impl.cc
+++ b/content/browser/web_contents/web_contents_impl.cc
@@ -10852,12 +10852,13 @@ void WebContentsImpl::NotifyFindReply(int request_id,
                                       int number_of_matches,
                                       const gfx::Rect& selection_rect,
                                       int active_match_ordinal,
-                                      bool final_update) {
+                                      bool final_update,
+                                      RenderFrameHost* target_rfh) {
   OPTIONAL_TRACE_EVENT0("content", "WebContentsImpl::NotifyFindReply");
   if (delegate_ && !IsBeingDestroyed() &&
       !GetPrimaryMainFrame()->GetProcess()->FastShutdownStarted()) {
     delegate_->FindReply(this, request_id, number_of_matches, selection_rect,
-                         active_match_ordinal, final_update);
+                         active_match_ordinal, final_update, target_rfh);
   }
 }
 
diff --git a/content/browser/web_contents/web_contents_impl.h b/content/browser/web_contents/web_contents_impl.h
index 55db63b3f99e465d2c77bb25dca5d50005df5ce7..0336af295da4404edfd20a48e7ff1f2b5fa290b5 100644
--- a/content/browser/web_contents/web_contents_impl.h
+++ b/content/browser/web_contents/web_contents_impl.h
@@ -1356,7 +1356,8 @@ class CONTENT_EXPORT WebContentsImpl
                        int number_of_matches,
                        const gfx::Rect& selection_rect,
                        int active_match_ordinal,
-                       bool final_update);
+                       bool final_update,
+                       RenderFrameHost* target_rfh);
 
   // Modify the counter of connected devices for this WebContents.
   void IncrementBluetoothConnectedDeviceCount();
diff --git a/content/public/browser/web_contents_delegate.h b/content/public/browser/web_contents_delegate.h
index 543c1f5a0dc04fdaaf7a332ec9869d286f736763..17d242c2d351891f1e68d581ff162fac35ee9d7a 100644
--- a/content/public/browser/web_contents_delegate.h
+++ b/content/public/browser/web_contents_delegate.h
@@ -585,7 +585,8 @@ class CONTENT_EXPORT WebContentsDelegate {
                          int number_of_matches,
                          const gfx::Rect& selection_rect,
                          int active_match_ordinal,
-                         bool final_update) {}
+                         bool final_update,
+                         RenderFrameHost* target_rfh) {}
 
 #if BUILDFLAG(IS_ANDROID)
   // Provides the rects of the current find-in-page matches.
diff --git a/extensions/browser/guest_view/web_view/web_view_guest.cc b/extensions/browser/guest_view/web_view/web_view_guest.cc
index 5adcd5d4c125b40088f938f0cfc2d21b0fe099ef..5c70a3546a44937e3214e0797254373a641d5722 100644
--- a/extensions/browser/guest_view/web_view/web_view_guest.cc
+++ b/extensions/browser/guest_view/web_view/web_view_guest.cc
@@ -772,9 +772,11 @@ void WebViewGuest::FindReply(WebContents* source,
                              int number_of_matches,
                              const gfx::Rect& selection_rect,
                              int active_match_ordinal,
-                             bool final_update) {
+                             bool final_update,
+                             content::RenderFrameHost* target_rfh) {
   GuestViewBase::FindReply(source, request_id, number_of_matches,
-                           selection_rect, active_match_ordinal, final_update);
+                           selection_rect, active_match_ordinal, final_update,
+                           target_rfh);
   find_helper_.FindReply(request_id, number_of_matches, selection_rect,
                          active_match_ordinal, final_update);
 }
diff --git a/extensions/browser/guest_view/web_view/web_view_guest.h b/extensions/browser/guest_view/web_view/web_view_guest.h
index 503cbecb388d0de471fc8dda3fc9688c4439a71e..5dd018e658865edbaa8d9789b8de87363d3d0f51 100644
--- a/extensions/browser/guest_view/web_view/web_view_guest.h
+++ b/extensions/browser/guest_view/web_view/web_view_guest.h
@@ -203,7 +203,8 @@ class WebViewGuest : public guest_view::GuestView<WebViewGuest> {
                  int number_of_matches,
                  const gfx::Rect& selection_rect,
                  int active_match_ordinal,
-                 bool final_update) final;
+                 bool final_update,
+                 content::RenderFrameHost* target_rfh) final;
   bool ZoomPropagatesFromEmbedderToGuest() const final;
   const char* GetAPINamespace() const final;
   int GetTaskPrefix() const final;
